### ISSUE TITLE
>5% top loci missing in genetics portal

### ISSUE THREAD

**Issue by d0choa** [opentargets/issues#2073]
*tl;dr*
*As of today, more than 5% of the top loci are not available in the genetics portal, due to a series of issues described in this ticket. The majority of absent loci are not the consequence of any recently introduced bug, but instead the result of some design choices in the architecture of Genetics Portal. This ticket does not aim to find or discuss solutions, just describe the causes of the issue to help with future strategy.*

It was previously reported (opentargets/issues#2344, opentargets/issues#2064, https://github.com/opentargets/issues/issues/2710) that some top loci signals went missing in the genetics portal. I decided to address this problem systematically and extract all top loci that are not present in the variant-to-disease ([V2D index](https://github.com/opentargets/genetics-pipe/blob/master/src/main/scala/ot/geckopipe/index/V2DIndex.scala)) to later analyse and try to explain the cause of this issue.

There is a total of 16,680 missing top loci corresponding to 5.6% of all top loci reported by the [V2D pipeline](https://github.com/opentargets/genetics-v2d-data). 

As a consequence, there are some inconsistencies in the Genetics Portal. For example, for the study `GCST90018757` the search menu reports 39 loci, but only 36 are available in the tables.

Next, I describe a list of the causes I identified that explain all the missing cases.

### Case 1: LD expansion

This is the most recurrent case. 15,668 top loci are not in the `ld` dataset (`gs://genetics-portal-dev-staging/v2d/220210/ld.parquet`). These are studies with no summary stats, therefore is relatively common for users to see them in the GWAS catalog, but are not available through Genetics Portal. 

```
 isLeadInVariantIndex  | true  
 isInLdOrFm            | false 
 isTagInVariantIndex   | false 
 isInStudyTable        | true  
 isInFINNGEN_CHD_STUDY | false 
 has_sumstats          | false 
 count(toplociId)      | 15668

```

For reference, this is an example of such top loci in the study [GCST90018757](https://genetics.opentargets.org/study/GCST90018757). 

```
 study_id              | GCST90018757        
 tag_chrom             | null                
 tag_pos               | null                
 tag_ref               | null                
 tag_alt               | null                
 lead_chrom            | X                   
 lead_pos              | 153627532           
 lead_ref              | C                   
 lead_alt              | T                   
 direction             | -                   
 beta                  | -0.0184248          
 beta_ci_lower         | -0.0243357810585671 
 beta_ci_upper         | -0.0125138189414328 
 odds_ratio            | null                
 oddsr_ci_lower        | null                
 oddsr_ci_upper        | null                
 pval_mantissa         | 1.0                 
 pval_exponent         | -9                  
 toplociId             | 8590140672          
 overall_r2            | null                
 pics_mu               | null                
 pics_postprob         | null                
 pics_95perc_credset   | null                
 pics_99perc_credset   | null                
 AFR_1000G_prop        | null                
 AMR_1000G_prop        | null                
 EAS_1000G_prop        | null                
 EUR_1000G_prop        | null                
 SAS_1000G_prop        | null                
 log10_ABF             | null                
 posterior_prob        | null                
 isInLdOrFm            | false               
 isLeadInVariantIndex  | true                
 isTagInVariantIndex   | false               
 has_sumstats          | false               
 isInStudyTable        | true                
 isInFINNGEN_CHD_STUDY | false               
```

### Case 2: Fine mapping
79 extra top loci are not available in `gs://genetics-portal-dev-staging/v2d/220210/finemapping.parquet`. These top loci are scattered in 21 different studies.

```
 isLeadInVariantIndex  | true  
 isInLdOrFm            | false 
 isTagInVariantIndex   | false 
 isInStudyTable        | true  
 isInFINNGEN_CHD_STUDY | false 
 has_sumstats          | true  
 count(toplociId)      | 75    

 isLeadInVariantIndex  | false 
 isInLdOrFm            | false 
 isTagInVariantIndex   | false 
 isInStudyTable        | true  
 isInFINNGEN_CHD_STUDY | false 
 has_sumstats          | true  
 count(toplociId)      | 4     
```

An example for future reference:
```
-RECORD 0-----------------------------------
 study_id              | GCST007800         
 tag_chrom             | null               
 tag_pos               | null               
 tag_ref               | null               
 tag_alt               | null               
 lead_chrom            | 1                  
 lead_pos              | 152215426          
 lead_ref              | T                  
 lead_alt              | C                  
 direction             | -                  
 beta                  | null               
 beta_ci_lower         | null               
 beta_ci_upper         | null               
 odds_ratio            | 0.5985728741957612 
 oddsr_ci_lower        | 0.5239195044098143 
 oddsr_ci_upper        | 0.6838636139850932 
 pval_mantissa         | 9.7                
 pval_exponent         | -23                
 toplociId             | 8590022899         
 overall_r2            | null               
 pics_mu               | null               
 pics_postprob         | null               
 pics_95perc_credset   | null               
 pics_99perc_credset   | null               
 AFR_1000G_prop        | null               
 AMR_1000G_prop        | null               
 EAS_1000G_prop        | null               
 EUR_1000G_prop        | null               
 SAS_1000G_prop        | null               
 log10_ABF             | null               
 posterior_prob        | null               
 isInLdOrFm            | false              
 isLeadInVariantIndex  | false              
 isTagInVariantIndex   | false              
 has_sumstats          | true               
 isInStudyTable        | true               
 isInFINNGEN_CHD_STUDY | false              
```
### Case 3: Variants not available in the variant index
All top loci are required to be represented in the variant index. At the moment, the variant index is a subset of GnomAD 2.x using a filter by minor allele frequency. 834 of the top loci are not represented in the variant index either as lead or tag variants.

```
 isLeadInVariantIndex  | false 
 isInLdOrFm            | true  
 isTagInVariantIndex   | false 
 isInStudyTable        | true  
 isInFINNGEN_CHD_STUDY | false 
 has_sumstats          | true  
 count(toplociId)      | 688   

 isLeadInVariantIndex  | true  
 isInLdOrFm            | true  
 isTagInVariantIndex   | false 
 isInStudyTable        | true  
 isInFINNGEN_CHD_STUDY | false 
 has_sumstats          | true  
 count(toplociId)      | 146   
```

The next variant is an example of this set. As expected, this variant is not available in GnomAD 2.x and it's also not observed in GnomAD 3.x. 

```
 study_id              | GCST90038680       
 tag_chrom             | 1                  
 tag_pos               | 152215426          
 tag_ref               | T                  
 tag_alt               | C                  
 lead_chrom            | 1                  
 lead_pos              | 152215426          
 lead_ref              | T                  
 lead_alt              | C                  
 direction             | +                  
 beta                  | null               
 beta_ci_lower         | null               
 beta_ci_upper         | null               
 odds_ratio            | 1.0283163981706804 
 oddsr_ci_lower        | 1.0244824231148917 
 oddsr_ci_upper        | 1.032164721315218  
 pval_mantissa         | 1.1                
 pval_exponent         | -48                
 toplociId             | 8590167270         
 overall_r2            | null               
 pics_mu               | null               
 pics_postprob         | null               
 pics_95perc_credset   | null               
 pics_99perc_credset   | null               
 AFR_1000G_prop        | null               
 AMR_1000G_prop        | null               
 EAS_1000G_prop        | null               
 EUR_1000G_prop        | null               
 SAS_1000G_prop        | null               
 log10_ABF             | 49.79451045888593  
 posterior_prob        | 0.250000000000001  
 isInLdOrFm            | true               
 isLeadInVariantIndex  | false              
 isTagInVariantIndex   | false              
 has_sumstats          | true               
 isInStudyTable        | true               
 isInFINNGEN_CHD_STUDY | false              
```

### Case 4: Missing from study table
There are 96 top loci missing because their study is missing from the table. All top loci belong to 2 single studies `GCST010653` and `GCST010729`. 

```
 isLeadInVariantIndex  | true  
 isInLdOrFm            | false 
 isTagInVariantIndex   | false 
 isInStudyTable        | false 
 isInFINNGEN_CHD_STUDY | false 
 has_sumstats          | null  
 count(toplociId)      | 96    
```

### Case 5: Problematic study in Finngen
I didn't find exactly the reason, but `FINNGEN_R5_I9_HEARTFAIL_AND_CHD` does not show in the V2D index. It contains 4 associated loci that are completely gone from production.

```
 isLeadInVariantIndex  | true  
 isInLdOrFm            | true  
 isTagInVariantIndex   | true  
 isInStudyTable        | true  
 isInFINNGEN_CHD_STUDY | true  
 has_sumstats          | true  
 count(toplociId)      | 3

 isLeadInVariantIndex  | true  
 isInLdOrFm            | true  
 isTagInVariantIndex   | false 
 isInStudyTable        | true  
 isInFINNGEN_CHD_STUDY | true  
 has_sumstats          | true  
 count(toplociId)      | 1     
```
Code to reproduce the diagnostics is available [here](https://gist.github.com/d0choa/b1162cf494bebcb033349888e2533fec)

There are some cases in which the lead variants have been dropped even before the top loci table. The next represent some possible cases:

### Case 6: Credible set lead vs top loci lead

As described in #2710, genetics portal currently pick the top loci for finngen [based on their posterior probability in the region](https://github.com/opentargets/genetics-finemapping/blob/ed7146900518562f4513b6e24749a9089412d039/finngen/finngen_finemapping_ingest.py#L70). In some ocassions, several credible sets can be derived from the same region. This case is more dramatic in the case of Sussie-derived fiine mapping (e.g. Finngen). 

For example, from the next 2 SNPs in the same region:

```
❯ gsutil cat 'gs://finngen-public-data-r6/finemapping/full/finngen_R6_I9_ANGINA.SUSIE.snp.bgz' | zcat | grep -P 'trait|19\:44908822\:C\:T|19\:44919689\:A\:G' | csvcut -d $'\t' -c 'trait,region,v,prob'
trait,region,v,prob
I9_ANGINA,chr19:43408822-46408822,19:44908822:C:T,0.999355892527265
I9_ANGINA,chr19:43408822-46408822,19:44919689:A:G,0.224710652837538
```

19:44908822:C:T was included as top loci for the trait and 19:44919689:A:G was discarded despite the variant being a lead in the credible set.

At the moment, this adds an inconsistency as credible sets are used for colocalisation analysis, so lead variants observed there might not reproduce in top loci tables.

**Comment by ireneisdoomed** [comment]
Case 5 has been fixed - more details in #2585

**Comment by tskir** [comment]
Probably related case reported through community website: https://community.opentargets.org/t/tlr7-in-open-targets-genetics/572. (TLR7 gene, present in GWAS Catalog: https://www.ebi.ac.uk/gwas/genes/TLR7)

**Comment by d0choa** [comment]
We are addressing this in https://github.com/opentargets/genetics_etl_python
